README.md
# fetcher

A simple fetcher CloudFlare worker implementation as a request delegation handler.



CHANGELOG.md
# Changelog

## 0.0.1 (2022-07-23)


### Features

* adds initial implementation ([b3bb12c](https://github.com/state303/fetcher/commit/b3bb12cd86cf37e8dca369402de2ef54b459690d))


### Miscellaneous Chores

* release 0.0.1 ([c28dd0c](https://github.com/state303/fetcher/commit/c28dd0cd59dea5bb9ec1546fa414938d3aff3dec))


test/index.test.ts
export {}

describe('dummy-test-suite', () => {
    it('dummy-test-run', () => {
        expect('hello'.concat(' ').concat('world')).toBe('hello world');
    });
});

test/handler.test.ts
import {
    handleRequest,
    MISSING_PROTOCOL_PREFIX,
    MISSING_TARGET_REQUEST_PARAMETER,
    UNSUPPORTED_METHOD
} from "../src/handler";

describe('handleRequestMethod test', () => {
    it('should allow GET method', (done) => {
        const req = new Request('https://www.test.com/', {method: 'GET'});
        handleRequest(req, fetch).then(response => {
            expect(response.status).toBe(400);
            expect(response.text().then(text => {
                expect(text).toContain(MISSING_TARGET_REQUEST_PARAMETER);
                done();
            }));
        });
    });

    it('should not allow POST method', (done) => {
        const req = new Request('https://www.test.com/', {method: 'POST'});
        handleRequest(req, fetch).then(response => {
            expect(response.status).toBe(405);
            expect(response.text().then(text => expect(text).toContain(UNSUPPORTED_METHOD)));
        });
        done();
    });

    it('should not accept empty target', (done) => {
        const req = new Request('https:/www.test.com/?target=', {method: 'GET'});
        handleRequest(req, fetch).then(response => {
            expect(response.status).toBe(400);
            expect(response.text().then(text => expect(text).toContain(MISSING_TARGET_REQUEST_PARAMETER)));
        });
        done();
    });

    it('should not accept malformed url', (done) => {
        const req = new Request('https://www.test.com/?target=https://:/#/')
        handleRequest(req, fetch).then(response => {
            expect(response.status).toBe(400);
            expect(response.text().then(text => {
                expect(text).toContain("://:/");
                done();
            }));
        })
    });

    it('should not accept GET method', (done) => {
        const req = new Request('https://www.test.com/?target=google.com', {method: 'GET'});
        handleRequest(req, fetch).then(response => {
            expect(response.status).toBe(400);
            expect(response.text().then(text => {
                    expect(text).toContain(MISSING_PROTOCOL_PREFIX);
                    done();
                }
            ));
        });
    });

    it('should pipe all data', (done) => {
        const req = new Request('https://www.test.com/?target=http://www.azenv.net', {method: 'GET'});

        async function mockFetch(request: string | Request, requestInitr?: Request | RequestInit | undefined): Promise<Response> {
            return new Response('test item', { status: 200 });
        }

        handleRequest(req, mockFetch).then(response => {
            expect(response.status).toBe(200);
            expect(response.text().then(text => {
                expect(text).toContain('test item');
                done();
            }));
        });
    });

    it('should delegate error response', (done) => {
        const req = new Request('https://www.test.com/?target=http://www.azenv.net', {method: 'GET'});

        async function mockFetch(request: string | Request, requestInitr?: Request | RequestInit | undefined): Promise<Response> {
            return new Response('not found', { status: 404 });
        }

        handleRequest(req, mockFetch).then(response => {
            expect(response.status).toBe(404);
            expect(response.text().then(text => {
                expect(text).toContain('not found');
                done();
            }));
        });
    });
});


src/index.ts
/**
 * Welcome to Cloudflare Workers! This is your first worker.
 *
 * - Run `wrangler dev src/index.ts` in your terminal to start a development server
 * - Open a browser tab at http://localhost:8787/ to see your worker in action
 * - Run `wrangler publish src/index.ts --name my-worker` to publish your worker
 *
 * Learn more at https://developers.cloudflare.com/workers/
 */
import {handleRequest} from "./handler";

export interface Env {
	// Example binding to KV. Learn more at https://developers.cloudflare.com/workers/runtime-apis/kv/
	// MY_KV_NAMESPACE: KVNamespace;
	//
	// Example binding to Durable Object. Learn more at https://developers.cloudflare.com/workers/runtime-apis/durable-objects/
	// MY_DURABLE_OBJECT: DurableObjectNamespace;
	//
	// Example binding to R2. Learn more at https://developers.cloudflare.com/workers/runtime-apis/r2/
	// MY_BUCKET: R2Bucket;
}

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext
	): Promise<Response> {
		return handleRequest(request, fetch);
	}
};




src/handler.ts
const PARAM = "target";
export const UNSUPPORTED_METHOD = "unsupported method";
export const MISSING_TARGET_REQUEST_PARAMETER = "missing target request parameter";
export const MISSING_PROTOCOL_PREFIX = "missing protocol prefix";
export const MALFORMED_URL = "malformed url";

export async function handleRequest(request: Request, fetch: (request: (Request | string), requestInitr?: (RequestInit | Request)) => Promise<Response>): Promise<Response> {
    if (request.method !== 'GET') {
        return new Response(`${UNSUPPORTED_METHOD}: ${request.method}`, {status: 405});
    }

    const target = new URL(request.url).searchParams.get(PARAM);

    if (target === null || target.length === 0) {
        return new Response(`${MISSING_TARGET_REQUEST_PARAMETER}`, {status: 400});
    }

    if (!target.startsWith("http://") && !target.startsWith("https://")) {
        return new Response(`${MISSING_PROTOCOL_PREFIX}: ${target}`, {status: 400});
    }

    try {
        new URL(target);
    } catch {
        return new Response(`${MALFORMED_URL}: ${target}`, {status: 400});
    }

    let {readable, writable} = new TransformStream();

    return fetch(target)
        .then(response => {
            response.body?.pipeTo(writable);
            return new Response(readable, response);
        });
}
